src/state/createStateProvider.tsx

```diff
@@ -0,0 +1,23 @@
+import React, { useContext } from 'react'
+
+/**
+ * This creates a "state provider".
+ *
+ * @param hookFn the hook function to return the context value
+ * @returns The context provider component and the useContextHook
+ */
+export function createStateProvider<Value>(hookFn: () => Value): [React.FunctionComponent<{ children: React.ReactNode }>, () => Value] {
+  const Context = React.createContext<Value | undefined>(undefined)
+  function WithContext({ children }: { children: React.ReactNode }) {
```

Comment:
You could put `interface Props { children: React.ReactNode }` at the top, and then you could do `React.FunctionComponent<Props>` and `({children}: Props)` instead of duplicating the props type.